### PR TITLE
1.19 support 

### PR DESCRIPTION
**Pull request by rom1504**

### CODE REVIEW

**Comment by rom1504** [comment]
did chat more or less

next:
checking https://github.com/PrismarineJS/minecraft-data/pull/587/commits/4caba63eea29c269dc0077a22fe8bf4aefbfe4c1 https://github.com/PrismarineJS/minecraft-data/pull/587/commits/4e845dcbdeba04b5a5e1cde66fd6914678da665a https://github.com/PrismarineJS/minecraft-data/pull/587/commits/f816c05200187ea71c15f2619caa242e2bd8cf1d https://github.com/PrismarineJS/minecraft-data/pull/587/commits/5ef06f1c54ee818a67ec2c5cd884b9c32fd2e1c6 to see what to add here

checking if any prismarine-* need change as well

**Comment by rom1504** [comment]
Error: [Prismarine-chunk] No chunk implementation for pc 1.19 found

**Comment by rom1504** [comment]
https://github.com/PrismarineJS/prismarine-chunk/pull/196

**Comment by rom1504** [comment]
pchunk done

**Comment by rom1504** [comment]
next handle all changes from https://wiki.vg/index.php?title=Pre-release_protocol&oldid=17659

immediate one is handle dimension -> registryCodec in login packet

**Comment by rom1504** [comment]
https://github.com/PrismarineJS/mineflayer/blob/9497b67d85082317bf7bf7a19427ed9371cb2f49/lib/plugins/game.js#L25 needs fixing

**Comment by extremeheat** [comment]
https://github.com/PrismarineJS/minecraft-data/pull/621 & https://github.com/PrismarineJS/prismarine-registry/pull/14 fix the login packet handling

**Comment by rom1504** [comment]
Uncaught TypeError: Cannot read property 'value' of undefined
      at Object.simplify (node_modules/prismarine-nbt/nbt.js:155:25)
      at handleRespawnPacketData (lib/plugins/game.js:40:33)
      at Client.<anonymous> (lib/plugins/game.js:58:5)


login packet handling need fixing there

**Comment by rom1504** [comment]
So there's dimension and dimensionCodec in login packet
https://github.com/PrismarineJS/mineflayer/blob/master/lib/plugins/game.js#L39

In 1.19 there's only dimensionCodec

Where is world max height located now ?

We need to fix that line of game.js

File: index.d.ts [hunk 1]
```diff
@@ -456,7 +456,10 @@ export interface Player {
   gamemode: number
   ping: number
   entity: Entity
-}
+  profileKeys?: {
+    publicKey: Buffer
+    signature: Buffer
+  }
```
Comment (OldStarchy) [inline index.d.ts:462]: missing }

**Comment by rom1504** [comment]
here needed is https://github.com/PrismarineJS/minecraft-data/issues/636

**Comment by extremeheat** [comment]
need https://github.com/PrismarineJS/prismarine-registry/pull/18
and fix release for https://github.com/PrismarineJS/prismarine-chat/actions/runs/3159953313/jobs/5143822604 

Also looks like spawn_entity packet structure may be incorrect
```
PartialReadError: Read error for undefined : undefined
    at new ExtendableError (/workspace/mineflayer/node_modules/protodef/src/utils.js:63:13)
    at new PartialReadError (/workspace/mineflayer/node_modules/protodef/src/utils.js:70:5)
    at Object.readVarInt [as varint] (/workspace/mineflayer/node_modules/protodef/src/datatypes/utils.js:69:45)
    at Object.packet_spawn_entity (eval at compile (/workspace/mineflayer/node_modules/protodef/src/compiler.js:258:12), <anonymous>:499:65)
    at eval (eval at compile (/workspace/mineflayer/node_modules/protodef/src/compiler.js:258:12), <anonymous>:2312:64)
```

**Comment by rom1504** [comment]
Chat seems to fail with these changes

**Comment by rom1504** [comment]
Tests are passing
Next steps:
* [ ] Check non tested examples still run

**Comment by rom1504** [comment]
Actually tests definitely not passing

File: lib/plugins/chat.js [hunk 2]
```diff
@@ -191,11 +219,61 @@ function inject (bot, options) {
...
-    const USERNAME_REGEX = '(?:\\(.{1,15}\\)|\\[.{1,15}\\]|.){0,5}?(\\w+)'
+    // 1.19 changes the chat format to move <sender> prefix from message contents to a seperate field.
+    // TODO: new chat lister to handle this
     bot.addChatPattern('whisper', new RegExp(`^${USERNAME_REGEX} whispers(?: to you)?:? (.*)$`), { deprecated: true })
     bot.addChatPattern('whisper', new RegExp(`^\\[${USERNAME_REGEX} -> \\w+\\s?\\] (.*)$`), { deprecated: true })
-    bot.addChatPattern('chat', new RegExp(`^${USERNAME_REGEX}\\s?[>:\\-»\\]\\)~]+\\s(.*)$`), { deprecated: true })
+    bot.addChatPattern('chat', LEGACY_VANILLA_CHAT_REGEX, { deprecated: true })
+  }
+
+  bot.awaitMessage2 = (opts) => {
```
Comment (rom1504) [inline lib/plugins/chat.js:233]: hmm why awaitMessage and awaitMessage2 ?

Comment (extremeheat) [reply lib/plugins/chat.js:233]: awaitMessage2 has more options and can match message content/sender without using regex on 1.19+

**Comment by rom1504** [comment]
failures left:
```
  4 failing

  1) mineflayer_external 1.19
       trade:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'polished_andesite'
- 'pumpkin_pie'
      + expected - actual

      -polished_andesite
      +pumpkin_pie
      
      at /home/runner/work/mineflayer/mineflayer/test/externalTests/trade.js:43:14
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  2) mineflayer_external 1.19
       "before each" hook for "useChests":
     Error: Timeout of 600000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/runner/work/mineflayer/mineflayer/test/externalTest.js)
      at listOnTimeout (internal/timers.js:557:17)
      at processTimers (internal/timers.js:500:7)

  3) mineflayer_internal 1.19
       world
         switchWorld respawn:
     Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/runner/work/mineflayer/mineflayer/test/internalTest.js)
      at listOnTimeout (internal/timers.js:557:17)
      at processTimers (internal/timers.js:500:7)

  4) mineflayer_internal 1.19
       "after each" hook for "switchWorld respawn":
     Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/runner/work/mineflayer/mineflayer/test/internalTest.js)
      at listOnTimeout (internal/timers.js:557:17)
      at processTimers (internal/timers.js:500:7)
```

**Comment by rom1504** [comment]
* trading external
* useChests external
* switch world internal

**Comment by extremeheat** [comment]
switchWorld is passing now with #2788 , rest are still failing

**Comment by IceTank** [comment]
Looks like entities test is failing because of the `spawn_entity_living` packet. The packet name is wrong/misleading. It should be named `named_entity_spawn`.  After a name change the entity tests should work again.

File: test/internalTest.js [hunk 3]
```diff
@@ -473,9 +505,10 @@ for (const supportedVersion of mineflayer.testedVersions) {
           // Versions prior to 1.11 have capital first letter
           const entities = bot.registry.entitiesByName
           const creeperId = entities.creeper ? entities.creeper.id : entities.Creeper.id
-          client.write('spawn_entity_living', {
+          client.write(bot.registry.supportFeature('consolidatedEntitySpawnPacket') ? 'spawn_entity' : 'spawn_entity_living', {
             entityId: 8, // random
             entityUUID: '00112233-4455-6677-8899-aabbccddeeff',
+            objectUUID: '00112233-4455-6677-8899-aabbccddeeff',
```
Comment (extremeheat) [inline test/internalTest.js:511]: In 1.19, `spawn_entity_living` was merged into `spawn_entity`

`spawn_entity` and `spawn_entity_living` are compatible with each other except for the fact that you need to write both entityUUID and objectUUID. To remove needing to fill out both fields we could replace that parameter with `uuid` on all versions

**Comment by extremeheat** [comment]
Fixed the issue with entity spawning above, with minor comment above about UUID field

Most tests minus villager one for 1.19 are passing with https://github.com/PrismarineJS/node-minecraft-data/pull/240

**Comment by rom1504** [comment]
1.13 fishing failing
1.19 use chest and trading failing

**Comment by IceTank** [comment]
Fishing for 1.13 is also failing? Was that working before?

**Comment by rom1504** [comment]
Yes all non 1.19 tests were working before

**Comment by extremeheat** [comment]
ok, everything is passing minus 1.13 fishing. Not clear why, nothing seems to have changed between 1.12 - 1.14

https://github.com/PrismarineJS/minecraft-data/blob/master/data/pc/1.12/entities.json#L3
https://github.com/PrismarineJS/minecraft-data/blob/master/data/pc/1.13.2/entities.json#L315
https://github.com/PrismarineJS/minecraft-data/blob/master/data/pc/1.14/entities.json#L345

It looks like https://github.com/PrismarineJS/mineflayer/blob/e28e878b3cdeac7040104461a088a29f93492011/lib/plugins/fishing.js#L24 is not getting called

**Comment by rom1504** [comment]
Could it be related to change made wrt spawn entity packet ?

**Comment by Shorent** [comment]
In 1.13.2 tests:
    test/externalTests/fishing.js, the variable "collected" has a name of item-stack.
    "item-stack" is not equal to "item". Therefore, the promise is never resolved.

**Comment by extremeheat** [comment]
yeah, I'm looking through the source code for 1.13 entities and the data in mc-data doesn't make sense. There's entities that aren't registered ("evocation_fangs") and the IDs are wrong.

**Comment by rom1504** [comment]
Maybe we can copy from 1.14 and remove additional? Or redo the extraction

**Comment by rom1504** [comment]
I looked at the changes and it looks good to me.

Next steps :
* Fixing that fishing test for 1.13
* testing manually inventory.js chest.js digger.js jumper.js

Is someone interested to running these examples now ? Would gain some time

**Comment by rom1504** [comment]
We'll need to support 1.19.2 next but I figure it'll be minor

**Comment by IceTank** [comment]
> By manual testing, do you mean testing the plug-ins in the game?

Those are examples for mineflayer usage in the examples folder. Manual testing in this case would just be trying those examples on a real server and see if they still function.

**Comment by Shorent** [comment]
Ok! I'll work on that 👍

**Comment by Shorent** [comment]
> > By manual testing, do you mean testing the plug-ins in the game?
> 
> Those are examples for mineflayer usage in the examples folder. Manual testing in this case would just be trying those examples on a real server and see if they still function.

I don't have much time for this today unfortunately, but would I be testing these examples for each supported version?

**Comment by rom1504** [comment]
> I don't have much time for this today unfortunately, but would I be testing these examples for each supported version?

no focus on 1.19

**Comment by rom1504** [comment]
all tests passing

**Comment by Shorent** [comment]
All tests fail with:
`Error: Unsupported brand channel name
    at getBrandCustomChannelName (/home/matthew/Documents/project/mc-bot/tests/19.9PR/mineflayer/lib/plugin/game.js:22:11)
    at inject (/home/matthew/Documents/project/mc-bot/tests/19.9PR/mineflayer/lib/plugins/game.js:67:24)
    at /home/matthew/Documents/project/mc-bot/tests/19.9PR/mineflayer/lib/plugin_loader.js:41:7
    at Array.forEach (<anonymous>)
    at injectPlugins (/home/matthew/Documents/project/mc-bot/tests/19.9PR/mineflayer/lib/plugin_loader.js:40:16)
    at EventEmitter.onInjectAllowed (/home/matthew/Documents/project/mc-bot/tests/19.9PR/mineflayer/lib/plugin_loader.js:12:5)
    at Object.onceWrapper (node:events:627:28)
    at EventEmitter.emit (node:events:513:28)
    at Client.next (/home/matthew/Documents/project/mc-bot/tests/19.9PR/mineflayer/lib/loader.js:124:9)
    at Object.onceWrapper (node:events:627:28)`

**Comment by Shorent** [comment]
jumper.js
=======
- Move Vehicle tests failed

I have tested inventory and digger, testing chest now

**Comment by rom1504** [comment]
> Move Vehicle tests failed

what information do you have ?

**Comment by rom1504** [comment]
I just tested inventory.js, works

**Comment by rom1504** [comment]
* [x] inventory.js
* [ ] chest.js
* [ ]  digger.js
* [ ] jumper.js

**Comment by rom1504** [comment]
testing jumper.js seeing some weird things regarding water

> jumper lost connection: Invalid move player packet received

when going into it

**Comment by rom1504** [comment]
pushed a fix for jumper.js vehicle selection

**Comment by rom1504** [comment]
tracking water issue in https://github.com/PrismarineJS/mineflayer/issues/2796

**Comment by rom1504** [comment]
I think this works well enough, let's go

File: index.d.ts [hunk 4]
```diff
@@ -456,7 +456,10 @@ export interface Player {
   gamemode: number
   ping: number
   entity: Entity
-}
```
Comment (floge07) [inline index.d.ts:459]: Uhm... 
image
And I wondered why my project won't compile...

**Pull request merged**